_source/_posts/2021-11-16-k8s-security-best-practices.md

```diff
@@ -0,0 +1,148 @@
+
+## 3. Use Secrets for all sensitive data with appropriate access
+
+This one should be a no-brainer. Kubernetes has a [Secret](https://kubernetes.io/docs/concepts/configuration/secret) resource that can be used to store sensitive data. This is a great way to store passwords, keys, and other sensitive data. Secrets can be used for storing string data, docker config, certificates, tokens, files, and so on. Secrets can be mounted as data volumes or exposed as environment variables to be used in containers. Secrets can be plain text or encoded, but please don't be that person who uses plain text secrets.
+
+Secrets are flexible and native to Kubernetes, so there is no reason for you not to use them. Also, make sure to implement proper RBAC for secrets so that not everyone has access to them.
+
+## 4. Keep Kubernetes version up to date
+
+Like any other software, Kubernetes also has bugs and issues. And from time to time, there might be a high severity bug that calls for a [CVE](https://en.wikipedia.org/wiki/Common_Vulnerabilities_and_Exposures). Hence, it's an excellent idea to keep the Kubernetes version up-to-date on the server and the CLI client. If you are using a managed PaaS, it should be pretty easy to upgrade, and for on-prem installations, there are tools like [kOps](https://kops.sigs.k8s.io/), [kubeadm](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/), and so on, that makes it easy to upgrade clusters.
+
+## 5. Restrict kubelet, API, and SSH access
+
+[kubelet](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/) is the primary "node agent" running on each node, and by default, a kubelet's HTTP endpoints are not secured. This could be an issue and hence [should be restricted](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet-authentication-authorization/).
+
+When someone has access to a Kubernetes cluster, they can access the k8s API server and SSH into the cluster nodes themselves. Ideally, this should be limited as much as possible. Disable SSH access to non-admin users. Secure the API server using OIDC and RBAC, as we saw earlier so that only authenticated users with sufficient roles can access the API.
+
+## 6. Secure container images
+
+Securing the container images that run on the cluster is as important as securing the cluster itself. A malicious image running on a cluster could wreak havoc. Follow these best practices for container image security.
+
+- Do not run containers as root as this would give the container unlimited access to the host. Always run the containers using a non-root user
```

Comment:
```suggestion
- Do not run containers as root as this would give the container unlimited access to the host. Always run the containers using a non-root user.
```